kernels/logcumsumexp/logcumsumexp.cpp

```diff
@@ -0,0 +1,160 @@
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS for A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS self.tcp LIABLE for ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *******************************************************************************/
+#include "logcumsumexp.h"
+#include <string>
+#include "core/context.h"
+#include "core/gen_case.h"
+#include "core/logging.h"
+#include "core/runtime/device.h"
+#include "core/tensor.h"
+#include "core/type.h"
+
+#define N_ALIGN 128
+#define CoreCapacity 327680  // memory length of input per core
+#define ClusterCapacity 1310720  // memory length of input per cluster
+#define DimOneDealLength 147456  // size of one NRAM in dim-one
+
+mluOpStatus_t MLUOP_WIN_API
+mluOpLogcumsumexp(mluOpHandle_t handle,
+                const int32_t dim,
+                const mluOpTensorDescriptor_t input_desc,
+                const void *input,
+                const mluOpTensorDescriptor_t output_desc,
+                void *output) {
+    const std::string API = "[mluOpLogcumsumexp]";
+
+    cnrtFunctionType_t k_type;
+    cnrtDim3_t k_dim;
+
+    PARAM_CHECK(API, handle != NULL);
+    PARAM_CHECK(API, input_desc != NULL);
+    PARAM_CHECK(API, input != NULL);
+    PARAM_CHECK(API, output_desc != NULL);
+    PARAM_CHECK(API, output != NULL);
+    PARAM_CHECK(API, input_desc->dtype == MLUOP_DTYPE_FLOAT ||
+                input_desc->dtype == MLUOP_DTYPE_HALF);
+    PARAM_CHECK(API, output_desc->dtype == MLUOP_DTYPE_FLOAT ||
+                output_desc->dtype == MLUOP_DTYPE_HALF);
+    PARAM_CHECK(API, input_desc->dim == output_desc->dim);
+    PARAM_CHECK(API, input_desc->layout == MLUOP_LAYOUT_ARRAY);
+    PARAM_CHECK(API, output_desc->layout == MLUOP_LAYOUT_ARRAY);
+
+
+    if (dim < (-1) * input_desc->dim) {
+        LOG(ERROR) << API
+                << " this negative dim is invalid. Received dim=["
+                << dim << "]";
+        return MLUOP_STATUS_BAD_PARAM;
+    }
+    if (dim >= input_desc->dim) {
+        LOG(ERROR) << API
+                << " dim beyonds the dimension of tensor. Received dim=["
+                << dim << "]";
+        return MLUOP_STATUS_BAD_PARAM;
+    }
+
+    int _dim = dim;
+    // preprocess for negative dim
+    if (_dim < 0) {
+        _dim += input_desc->dim;
+    }

```

Comment:
76~ 80 这几句可以改成这样吧？
```suggestion
const int32_t dim_vaild = dim < 0 ? input_desc->dim : dim;
```
后续用 dim_vaild 这个参数就行，参数名 " dim_vaild " 只是随意构造，不属于建议

Reply:
然后63~73 的dim check 可以缩减为一个if判断

Reply:
收到